packages/common/src/cis2/util.ts

```diff
@@ -0,0 +1,504 @@
+        update:
+            | { Add: Record<string, never> }
+            | { Remove: Record<string, never> };
+        operator: AddressParamJson;
+    };
+}
+
+/**
+ * Checks whether an `Address` is a `ContractAddress`
+ */
+export const isContractAddress = (
+    address: CIS2.Address
+): address is ContractAddress => typeof address !== 'string';
+
+function serializeCIS2TokenId(tokenId: HexString): Buffer {
+    const serialized = Buffer.from(tokenId, 'hex');
+
+    if (serialized.length > TOKEN_ID_MAX_LENGTH) {
+        throw new Error(
+            `Token ID exceeds maximum size of ${TOKEN_ID_MAX_LENGTH} bytes`
+        );
+    }
+
+    return packBufferWithWord8Length(serialized);
+}
+
+function serializeTokenAmount(amount: bigint): Buffer {
+    const serialized = uleb128.encode(amount.toString());
+
+    if (serialized.length > TOKEN_AMOUNT_MAX_LENGTH) {
+        throw new Error(
+            `Token amount exceeds maximum size of ${TOKEN_AMOUNT_MAX_LENGTH} bytes`
+        );
+    }
+
+    return serialized;
+}
+
+function serializeAccountAddress(address: HexString): Buffer {
+    return new AccountAddress(address).decodedAddress;
+}
+
+function serializeContractAddress(address: ContractAddress): Buffer {
+    const index = encodeWord64(address.index, true);
+    const subindex = encodeWord64(address.subindex, true);
+    return Buffer.concat([index, subindex]);
+}
+
+function serializeAddress(address: CIS2.Address): Buffer {
+    const isContract = isContractAddress(address);
+    const type = encodeWord8(isContract ? 1 : 0);
+    const serializedAddress = !isContract
+        ? serializeAccountAddress(address)
+        : serializeContractAddress(address);
+
+    return Buffer.concat([type, serializedAddress]);
+}
+
+function serializeReceiveHookName(hook: string): Buffer {
+    const serialized = Buffer.from(hook, 'ascii');
+
+    if (serialized.length > TOKEN_RECEIVE_HOOK_MAX_LENGTH) {
+        throw new Error(
+            `Token receive hook name exceeds maximum size of ${TOKEN_RECEIVE_HOOK_MAX_LENGTH} bytes`
+        );
+    }
+
+    return packBufferWithWord16Length(serialized, true);
+}
+
+function serializeContractReceiver(receiver: CIS2.ContractReceiver): Buffer {
+    const address = serializeContractAddress(receiver.address);
+    const hook = serializeReceiveHookName(receiver.hookName);
+    return Buffer.concat([address, hook]);
+}
+
+function serializeReceiver(receiver: CIS2.Receiver): Buffer {
+    const isAccount = typeof receiver === 'string';
+    const type = encodeWord8(isAccount ? 0 : 1);
+    const serializedAddress = isAccount
+        ? serializeAccountAddress(receiver)
+        : serializeContractReceiver(receiver);
+
+    return Buffer.concat([type, serializedAddress]);
+}
+
+function serializeAdditionalData(data: HexString): Buffer {
+    const serialized = Buffer.from(data, 'hex');
+    return packBufferWithWord16Length(serialized, true);
+}
+
+const makeSerializeList =
```

Comment:
Consider following the style of the other function definitions and not use arrow-functions here

Reply:
I disagree here, as arrow functions are much more readable when you have to do anything functional, e.g. curried functions. I'll change arrow functions to function definitions where appropriate.

Reply:
Sure, if the style is to do this for curried functions. I am fine with it